[AIRFLOW-7068] Create EC2 Hook, Operator and Sensor

mustafagok opened the pull request:
Create EC2 Hook, Operator and Sensor

- New hook to interact with AWS EC2 Service.
- New operator to manage AWS EC2 instance and to change instance state by applying given operation using boto3.
- New sensor to check the state of the AWS EC2 instance until state of the instance become equal to the target state.

---
Issue link: [AIRFLOW-7068](https://issues.apache.org/jira/browse/AIRFLOW-7068)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

mustafagok committed:
[AIRFLOW-7068] Create EC2Hook

To interact with AWS EC2 Service.

mustafagok committed:
[AIRFLOW-7068] Create EC2Operator

Manage AWS EC2 instance using boto3.
Change instance state by applying given operation.

mustafagok committed:
[AIRFLOW-7068] Create EC2Sensor

Check the state of the AWS EC2 instance until
state of the instance become equal to the target state.

mustafagok committed:
[AIRFLOW-7068] Add EC2Hook, EC2Operator, EC2Sensor to docs

Move ECSOperator from EC2 to ECS part

mik-laj commented on airflow/providers/amazon/aws/operators/ec2.py:
```diff
@@ -0,0 +1,102 @@
+
+        instance = self.hook.get_conn().Instance(id=self.instance_id)
+
+        if self.operation == "start":
```
It is recommended to create one operator for each operation.

mustafagok replied:
I can write EC2StartInstanceOperator and EC2StopInstanceOperator, but these operators will be identical except 2 lines. What do you suggest?

turbaszek replied:
I agree with @mik-laj, we should two operators even if the difference is small :)

mik-laj commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -0,0 +1,48 @@
+        # set self.conn in first call
+        if not self.conn:
+            self.conn = self.get_resource_type("ec2", self.region_name)
+        return self.conn
```
This operator has too few methods. In an ideal world, operators should never call the get_conn method directly.

mustafagok replied:
You are right, I think I should add a method named get_instance(instance_id) or get_instance_state(instance_id) to this EC2Hook class. Then I can call this from operator/sensor. What do you think?

ashb commented on airflow/providers/amazon/aws/operators/ec2.py:
```diff
@@ -0,0 +1,102 @@
+        self.region_name = region_name
+        self.check_interval = check_interval
+        self.target_state = self.operation_target_state_map[self.operation]
+        self.hook = self.get_hook()
```
Do not create hooks in Operator constructors -- this could open a network connection before the task is parsed (i.e. at DAG parse time).

mustafagok replied:
fixed

ashb commented on airflow/providers/amazon/aws/operators/ec2.py:
```diff
@@ -0,0 +1,102 @@
+from airflow.utils.decorators import apply_defaults
+
+
+class EC2Operator(BaseOperator):
```
```suggestion
class EC2InstanceOperator(BaseOperator):
```

There are more than just Instances in EC2

mustafagok replied:
I have split it into 2 operators: EC2StartInstanceOperator and EC2StopInstanceOperator

ashb commented on airflow/providers/amazon/aws/sensors/ec2.py:
```diff
@@ -0,0 +1,78 @@
+    template_fields = ["target_state", "region_name"]
+    ui_color = "#cc8811"
+    ui_fgcolor = "#ffffff"
+    valid_states = ["running", "stopped"]
```
What about terminated?

mustafagok replied:
added, thx

ashb commented on airflow/providers/amazon/aws/sensors/ec2.py:
```diff
@@ -0,0 +1,78 @@
+        self.instance_id = instance_id
+        self.aws_conn_id = aws_conn_id
+        self.region_name = region_name
+        self.hook = self.get_hook()
```
Same here.

ashb replied:
The other problem with calling this here: the region_name template _will not_ have been rendered yet. That only happens just before execute is called.

mustafagok replied:
I didn't know that, fixed, thank you. 
P.S.: There are some bad practices in other operators.

ashb commented on airflow/providers/amazon/aws/operators/ec2.py:
```diff
@@ -0,0 +1,102 @@
+                 *args,
+                 **kwargs):
+        if operation not in self.valid_operations:
+            raise AirflowException(f"Invalid operation: {operation}")
```
```suggestion
            raise ValueError(f"Invalid operation: {operation}")
```

> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.

mustafagok replied:
This part removed, after spliting EC2StartInstanceOperator and EC2StopInstanceOperator

mustafagok committed:
[AIRFLOW-7068] Add methods to EC2Hook

get_instance: Get EC2 instance by id and return it.
get_instance_state: Get EC2 instance state by id and return it.

mustafagok committed:
[AIRFLOW-7068] Create EC2StartInstanceOperator, EC2StopInstanceOperator

It is recommended to create one operator for each operation.
So, split EC2Operator to 2 operators: EC2StartInstanceOperator,
EC2StopInstanceOperator

mustafagok committed:
[AIRFLOW-7068] Do not create hooks in constructors

This could open a network connection before the task is parsed

mustafagok committed:
[AIRFLOW-7068] Add terminated state to EC2Sensor

mustafagok committed:
[AIRFLOW-7068] Rename EC2Sensor as EC2InstanceStateSensor

turbaszek commented on airflow/providers/amazon/aws/operators/ec2_start_instance.py:
```diff
@@ -0,0 +1,88 @@
+            time.sleep(self.check_interval)
+            instance_state = ec2_hook.get_instance_state(
+                instance_id=self.instance_id
+            )
```
This method seems to be used in both operators. Can we extract it to hook as something like `wait_for_status` ?

mustafagok replied:
Sure, I will.

turbaszek commented on airflow/providers/amazon/aws/operators/ec2_start_instance.py:
```diff
@@ -0,0 +1,88 @@
+        return EC2Hook(
+            aws_conn_id=self.aws_conn_id,
+            region_name=self.region_name
+        )
```
Personally I would remove this method and just done:
```python
hook = EC2Hook(
             aws_conn_id=self.aws_conn_id,
             region_name=self.region_name
         )
```
in `execute` method :)

mustafagok replied:
You're right, a little unnecessary. I saw examples like this and thought "I should write like this". I will change it.

mustafagok committed:
[AIRFLOW-7068] Use ValueError i/o AirflowException

mustafagok committed:
[AIRFLOW-7068] Add wait_for_state method to EC2Hook

mustafagok committed:
[AIRFLOW-7068] Remove get_hook methods

mustafagok committed:
[AIRFLOW-7068] Add test_terminated unittest

feluelle commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -0,0 +1,98 @@
+        :return: None
+        :rtype: None
+        """
+        instance_state = self.get_instance_state(
+            instance_id=instance_id
+        )
+        while instance_state != target_state:
+            self.log.info("instance state: %s", instance_state)
+            time.sleep(check_interval)
+            instance_state = self.get_instance_state(
+                instance_id=instance_id
+            )
```
```suggestion
        while True:
            instance_state = self.get_instance_state(
                instance_id=instance_id
            )
            self.log.info("instance state: %s", instance_state)
            if instance_state == target_state: 
                break
            time.sleep(check_interval)
```
WDYT?

mustafagok replied:
Not sure, I heard that while True is a bad practice since it's hard to follow the loop condition, but the other one has some code duplication. So, I am not sure, what do you think about this, do you know any post related to this situation. I am facing this situation frequently and still not know what to do :(
[example question frrom stackoverflow](https://stackoverflow.com/questions/743164/emulate-a-do-while-loop-in-python)

feluelle replied:
I think both options are fine (in this case) but my pref would be the second. You can keep yours. LGTM too :)

feluelle commented on airflow/providers/amazon/aws/sensors/ec2_instance_state.py:
```diff
@@ -0,0 +1,77 @@
+        ec2_hook = self.get_hook()
+        instance = ec2_hook.get_instance(instance_id=self.instance_id)
+        self.log.info("instance state: %s", instance.state)
+        return instance.state["Name"] == self.target_state
```
```suggestion
        instance_state = ec2_hook.get_instance_state(instance_id=self.instance_id)
        self.log.info("instance state: %s", instance_state)
        return instance_state  == self.target_state
```

feluelle commented on docs/operators-and-hooks-ref.rst:
```diff
@@ -360,12 +360,19 @@ These integrations allow you to perform various operations within the Amazon Web
+     - :mod:`airflow.providers.amazon.aws.operators.ec2_start_instance`,
+       :mod:`airflow.providers.amazon.aws.operators.ec2_stop_instance`,
+     - :mod:`airflow.providers.amazon.aws.sensors.ec2_instance_state`
+
```
Can you put this block before `Amazon ECS`, please?

mustafagok replied:
Sure, I did not noticed that they are alphabetical

mustafagok committed:
Use get_instance_state i/o get_instance

As suggested by feluelle

Co-Authored-By: Felix Uellendall <feluelle@users.noreply.github.com>

mustafagok committed:
[AIRFLOW-7068] Move Amazon EC2 block before Amazon ECS

mustafagok committed:
[AIRFLOW-7068] Fix flake8 errors

mustafagok committed:
[AIRFLOW-7068] Make EC2Hook compatible with AwsBaseHook

See Pull Request #7541

mustafagok committed:
[AIRFLOW-7068] Fix flake8 errors

ashb commented on airflow/providers/amazon/aws/hooks/ec2.py:
```diff
@@ -0,0 +1,88 @@
+        :return: Instance object
+        :rtype: ec2.Instance
+        """
+        return self.get_conn().Instance(id=instance_id)
```
```suggestion
        return self.conn.Instance(id=instance_id)
```

mustafagok replied:
Oh, I have changed all other get_conn() methods with self.conn, but forgot this one :)
thanks

mustafagok committed:
Use self.conn i/o self.get_conn method

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>

ashb commented on airflow/providers/amazon/aws/operators/ec2_start_instance.py:
```diff
@@ -0,0 +1,72 @@
+    :type check_interval: float
+    """
+
+    template_fields = ["region_name"]
```
```suggestion
    template_fields = ("instance_id", "region_name")
```

mustafagok replied:
all done :heavy_check_mark: waiting travis to build

ashb commented on airflow/providers/amazon/aws/operators/ec2_stop_instance.py:
```diff
@@ -0,0 +1,72 @@
+    :type check_interval: float
+    """
+
+    template_fields = ["region_name"]
```
```suggestion
    template_fields = ("instance_id", "region_name")
```

ashb commented on airflow/providers/amazon/aws/sensors/ec2_instance_state.py:
```diff
@@ -0,0 +1,70 @@
+    :type region_name: Optional[str]
+    """
+
+    template_fields = ["target_state", "region_name"]
```
```suggestion
    template_fields = ("instance_id", "target_state", "region_name")
```

ashb commented on tests/providers/amazon/aws/sensors/test_ec2_instance_state.py:
```diff
@@ -0,0 +1,132 @@
+
+
+if __name__ == '__main__':
+    unittest.main()
```
```suggestion
```

ashb commented on tests/providers/amazon/aws/operators/test_ec2_stop_instance.py:
```diff
@@ -0,0 +1,68 @@
+
+
+if __name__ == '__main__':
+    unittest.main()
```
```suggestion
```

ashb commented on tests/providers/amazon/aws/operators/test_ec2_start_instance.py:
```diff
@@ -0,0 +1,68 @@
+
+
+if __name__ == '__main__':
+    unittest.main()
```
```suggestion
```

ashb commented on tests/providers/amazon/aws/hooks/test_ec2.py:
```diff
@@ -0,0 +1,75 @@
+
+
+if __name__ == '__main__':
+    unittest.main()
```
```suggestion
```

ashb reviewed (dismissed):
Couple of small changes, looking good.

Could you also rebase on to latest master too please?

mustafagok committed:
[AIRFLOW-7068] Add instance_id to template_fields

Remove unittest.main() from tests

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>

ashb commented on airflow/providers/amazon/aws/operators/ec2_start_instance.py:
```diff
@@ -0,0 +1,72 @@
+
+    template_fields = ("instance_id", "region_name")
+    ui_color = "#eeaa11"
+    ui_fgcolor = "#ffffff"
```
White as the FG colour? Can you show this in a screenshot please?

(Most of the other ones are black FG aren't they?)

mustafagok replied:
![image](https://user-images.githubusercontent.com/18196182/81109578-f9c02280-8f22-11ea-8f9c-48bc204be170.png)

I can change it

mustafagok replied:
maybe black could be better option for ec2 operator, but ec2 state sensor has darker background

feluelle replied:
LGTM

feluelle reviewed (changes requested):
Oh I missed that: Can you rename `EC2` to `Ec2`? Like we did for `Emr` or `Aws` for example?!
